driver-core/src/main/com/mongodb/operation/MixedBulkWriteOperation.java

```diff
@@ -183,19 +183,18 @@ public Boolean getRetryWrites() {
      */
     @Override
     public BulkWriteResult execute(final WriteBinding binding) {
-        return withConnection(binding, new CallableWithConnection<BulkWriteResult>() {
+        BulkWriteBatch bulkWriteBatch = withConnection(binding, new CallableWithConnection<BulkWriteBatch>() {
             @Override
-            public BulkWriteResult call(final Connection connection) {
+            public BulkWriteBatch call(final Connection connection) {
                 validateWriteRequests(connection, bypassDocumentValidation, writeRequests, writeConcern, retryWrites);
                 if (getWriteConcern().isAcknowledged() || serverIsAtLeastVersionThreeDotSix(connection.getDescription())) {
-                    BulkWriteBatch batch = BulkWriteBatch.createBulkWriteBatch(namespace, connection.getDescription(), ordered,
+                    return BulkWriteBatch.createBulkWriteBatch(namespace, connection.getDescription(), ordered,
                             writeConcern, bypassDocumentValidation, retryWrites, writeRequests, binding.getSessionContext());
-                    return executeBatches(binding, connection, batch, retryWrites);
-                } else {
-                    return executeLegacyBatches(connection);
                 }
+                return null;
             }
         });
+        return (bulkWriteBatch != null) ? executeBulkWriteBatch(binding, bulkWriteBatch) : executeLegacyBatches(binding);
```

Comment:
The driver shouldn't have to get and return a connection to the pool just to figure out the server version, or for each batch.  Rather it should get a connection and use it for the entire operation, only switching as required by a retryable failure.

Reply:
Done.